From 94bfcb70b5172ff72f3867085b3429b35b42a136 Mon Sep 17 00:00:00 2001
From: Portisch <hugo.portisch@yahoo.de>
Date: Tue, 6 Apr 2021 08:26:06 +0200
Subject: [PATCH] AmlogicGrabber: set and decrease frame grab timeout to 500ms
 This prevent an internal timeout when no frame is delivered like on playback
 pause.

---
 libsrc/grabber/amlogic/AmlogicGrabber.cpp | 3 ++-
 libsrc/grabber/amlogic/Amvideocap.h       | 1 +
 2 files changed, 3 insertions(+), 1 deletion(-)

diff --git a/libsrc/grabber/amlogic/AmlogicGrabber.cpp b/libsrc/grabber/amlogic/AmlogicGrabber.cpp
index 34d531d0..d4ee2f08 100644
--- a/libsrc/grabber/amlogic/AmlogicGrabber.cpp
+++ b/libsrc/grabber/amlogic/AmlogicGrabber.cpp
@@ -132,8 +132,9 @@ int AmlogicGrabber::grabFrame_amvideocap(Image<ColorRgb> & image)
 		long r1 = ioctl(_captureDev, AMVIDEOCAP_IOW_SET_WANTFRAME_WIDTH, _width);
 		long r2 = ioctl(_captureDev, AMVIDEOCAP_IOW_SET_WANTFRAME_HEIGHT, _height);
 		long r3 = ioctl(_captureDev, AMVIDEOCAP_IOW_SET_WANTFRAME_AT_FLAGS, CAP_FLAG_AT_END);
+		long r4 = ioctl(_captureDev, AMVIDEOCAP_IOW_SET_WANTFRAME_WAIT_MAX_MS, 500);
 
-		if (r1<0 || r2<0 || r3<0 || _height==0 || _width==0)
+		if (r1<0 || r2<0 || r3<0 || r4<0 || _height==0 || _width==0)
 		{
 			ErrorIf(_lastError != 2,_log,"Failed to configure capture device (%d - %s)", errno, strerror(errno));
 			_lastError = 2;
diff --git a/libsrc/grabber/amlogic/Amvideocap.h b/libsrc/grabber/amlogic/Amvideocap.h
index 828aa895..e38853d7 100644
--- a/libsrc/grabber/amlogic/Amvideocap.h
+++ b/libsrc/grabber/amlogic/Amvideocap.h
@@ -14,6 +14,7 @@
 // #define AMVIDEOCAP_IOW_SET_WANTFRAME_FORMAT     		_IOW(AMVIDEOCAP_IOC_MAGIC, 0x01, int)
 #define AMVIDEOCAP_IOW_SET_WANTFRAME_WIDTH      		_IOW(AMVIDEOCAP_IOC_MAGIC, 0x02, int)
 #define AMVIDEOCAP_IOW_SET_WANTFRAME_HEIGHT     		_IOW(AMVIDEOCAP_IOC_MAGIC, 0x03, int)
+#define AMVIDEOCAP_IOW_SET_WANTFRAME_WAIT_MAX_MS     		_IOW(AMVIDEOCAP_IOC_MAGIC, 0x05, unsigned long long)
 #define AMVIDEOCAP_IOW_SET_WANTFRAME_AT_FLAGS     		_IOW(AMVIDEOCAP_IOC_MAGIC, 0x06, int)
 
 #define _A_M  'S'
-- 
2.31.0

